test/common/upstream/cds_speed_test.cc

```diff
@@ -0,0 +1,191 @@
+                       "");
+        resource->set_type_url("type.googleapis.com/envoy.api.v2.ClusterLoadAssignment");
+      }
+    }
+    validation_visitor_.setSkipValidation(ignore_unknown_dynamic_fields);
+
+    state_.SetComplexityN(num_clusters);
+    state_.ResumeTiming();
+    grpc_mux_->grpcStreamForTest().onReceiveMessage(std::move(response));
+  }
+
+  TestDeprecatedV2Api _deprecated_v2_api_;
+  State& state_;
+  const bool v2_config_;
+  const std::string type_url_;
+  uint64_t version_{};
+  bool initialized_{};
+  Stats::IsolatedStoreImpl stats_;
+  Config::SubscriptionStats subscription_stats_;
+  Ssl::MockContextManager ssl_context_manager_;
+  envoy::config::cluster::v3::Cluster eds_cluster_;
+  NiceMock<MockClusterManager> cm_;
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  EdsClusterImplSharedPtr cluster_;
+  Config::SubscriptionCallbacks* eds_callbacks_{};
+  Config::OpaqueResourceDecoderImpl<envoy::config::endpoint::v3::ClusterLoadAssignment>
+      resource_decoder_{validation_visitor_, "cluster_name"};
+  NiceMock<Random::MockRandomGenerator> random_;
+  NiceMock<Runtime::MockLoader> runtime_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  NiceMock<Server::MockAdmin> admin_;
+  Singleton::ManagerImpl singleton_manager_{Thread::threadFactoryForTest()};
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  ProtobufMessage::MockValidationVisitor validation_visitor_;
+  Api::ApiPtr api_;
+  Grpc::MockAsyncClient* async_client_;
+  NiceMock<Grpc::MockAsyncStream> async_stream_;
+  Config::GrpcMuxImplSharedPtr grpc_mux_;
+  Config::GrpcSubscriptionImplPtr subscription_;
+};
+
+} // namespace Upstream
+} // namespace Envoy
+
+static void addClusters(State& state) {
+  Envoy::Thread::MutexBasicLockable lock;
+  Envoy::Logger::Context logging_state(spdlog::level::warn,
+                                       Envoy::Logger::Logger::DEFAULT_LOG_FORMAT, lock, false);
+  for (auto _ : state) { // NOLINT(clang-analyzer-deadcode.DeadStores)
+    Envoy::Upstream::CdsSpeedTest speed_test(state, state.range(0));
+    // if we've been instructed to skip tests, only run once no matter the argument:
+    uint32_t clusters = skipExpensiveBenchmarks() ? 1 : state.range(2);
+    speed_test.clusterHelper(state.range(1), clusters);
+  }
+}
+
+BENCHMARK(addClusters)
+    ->Ranges({{false, true}, {false, true}, {64, 100000}})
+    ->Unit(benchmark::kMillisecond)
+    ->Complexity();
+
+static void duplicateUpdate(State& state) {
+  Envoy::Thread::MutexBasicLockable lock;
+  Envoy::Logger::Context logging_state(spdlog::level::warn,
+                                       Envoy::Logger::Logger::DEFAULT_LOG_FORMAT, lock, false);
+
+  for (auto _ : state) { // NOLINT(clang-analyzer-deadcode.DeadStores)
+    Envoy::Upstream::CdsSpeedTest speed_test(state, false);
+    uint32_t clusters = skipExpensiveBenchmarks() ? 1 : state.range(0);
+
+    speed_test.clusterHelper(true, clusters);
```

Comment:
can you comment a bit on a few choices here:

1. Why call `speed_test.clusterHelper(true, clusters);` twice?
2. Why not construct `CdsSpeedTest` outside the loop?
3. Related: should we put the `Envoy::Logger::Context` into the class to avoid repeating that code for each test?

I'm not pushing back against those decisions I'm just trying to understand them.

Reply:
1) that's the difference between the two test cases; I added a comment (hopefully) explaining.
2) I assume the loop here is modifying state in a meaningful way, which is then passed to the newly constructed speed_test. This ```auto _ : state``` is directly from the example in https://github.com/google/benchmark
3) y'know, I don't know what the Logger Context was doing here, and removing it has no ill effect...

This was mostly copy & pasted from eds_speed_test (I don't understand github's copy detection and why it didn't notice this); I've backported the changes to eds_speed_test as well.

Reply:
RE 1 and 3: cool, all set, thanks.

Re 2: the `auto _ state : state` pattern is fine, but I think state.range(x) won't change within a call, and it would be  safe to construct the CdsSpeedTest object outside the loop. I think that will result in a more focused perf benchmark. WDYT?

Reply:
Testing bears out your hypothesis, I see what I missed there.